### PR TITLE
Executor_pool docs

### PR DESCRIPTION
**Pull request by SGrondin**
Changes to the README:
- Replaced a few `"5.0"` with `"5.1"` since Eio now requires OCaml 5.1
- Moved the `Domainslib` and `kcas` sections to their own files in `doc/`
- Rewrote the `Multicore Support` section to include, in order:
  - an overview of the tools available
  - a comparison between `Executor_pool` and `Domain_manager.run`. Assumption: expert users will know when they need to directly create threads, no need to explain the rationale to them.
  - docs and example for `Executor_pool`
  - docs and example for `Domain_manager.run`
  - a basic "Multicore Performance Guide" based on our conversation with @samoht
- ~I've moved the following sections to be after `Synchronization tools`:~
  - ~`Multicore Support` (the user shoul learn about Promise, Mutexes, Streams, etc. first)~
  - ~`Running processes` (same, and I expect this feature to see less frequent use than most modules that were placed ahead of it in the docs)~
  - ~`Tracing` (this is a somewhat advanced feature, mostly geared towards debugging. I find it easier to not constantly mix "features" with "testing+debugging" in the docs)~
  - ~`Testing with Mocks` (same as Tracing)~
- ~Moved the `Mock Clocks` section from `Time` to `Testing with Mocks`~
- Removed the `Worker Pool` example from the `Streams` section
  - I kept as much information as possible. Feel free to re-add anything here.

Please let me know if any of the above causes issues. Feel free to make changes directly on this branch.

### CODE REVIEW

File: README.md
Comment (Sudha247) [inline README.md:5]: Should we make this OCaml 5.1+?

File: README.md
Comment (Sudha247) [inline README.md:1283]: I'm not sure we need to mention "thread" here. It can be counter-intuitive. I believe thread here is referring to a thread of execution. In general a thread in OCaml means systhread by default.

File: README.md
Comment (Sudha247) [inline README.md:1272]: Maybe it's worth mentioning `weight` allows running computations in a domain until it reaches capacity. I understood this after looking at the mli file.

I'm also curious whether we can automate this somehow. There's a risk of users providing inaccurate weights. But that's not a conversation for this PR.

File: README.md
Comment (Sudha247) [inline README.md:1364]: Perhaps move the Domain Manager part on top and then introduce the Executor pool as an improvement?

File: README.md
Comment (Sudha247) [inline README.md:1416]: I think some of the issues with idle domains and performance were fixed in 5.1 https://github.com/ocaml/ocaml/pull/11903. It's ok to leave the note as is. If there are more issues, it might be worth reporting upstream.

**Review by Sudha247 — COMMENTED** [review]
Thanks @SGrondin! This is great. I've left some comments below.

File: README.md
Comment (balat) [reply README.md:5]: Or just OCaml 5

File: README.md
Comment (SGrondin) [reply README.md:1283]: I thought a lot about this yet I'm still unsure about it, so I appreciate your perspective.

The reasoning was that _I think_ that starting within the next 6 months, Eio will become the single most common entrypoint for OCaml 4 developers to learn about OCaml 5's multicore feature. Within the same timeframe, I also believe that Eio will be **the** place where C#, Java, Go, etc., developers learn how to perform I/O and use multithreading in OCaml. Those developers will be familiar with the concept of threads and I wanted to emphasize that whenever they see the word "domain", they can rest assured that it maps 1:1 to a thread.

Thoughts? What do you think?

File: README.md
Comment (SGrondin) [reply README.md:1364]: I'd prefer to keep the docs on the Executor Pool self-sufficient so users don't have to learn about a different module to understand it. I've made small tweaks to help with the confusion you've encountered.

File: README.md
Comment (SGrondin) [reply README.md:1272]: Good point, I've made a few changes to help.

I think it's okay if the average user ends up looking at the mli through their code editor or on the https://ocaml-multicore.github.io/eio/eio/Eio/ site once they start using a module if it helps keep the README more focused on just: the purpose of a module, how it fits within the library, and advice about using it effectively.

> I'm also curious whether we can automate this somehow. There's a risk of users providing inaccurate weights. But that's not a conversation for this PR.

I've been thinking the same thing but I haven't been able to figure out a decent abstraction that brings enough value to justify losing the simplicity (in the API) of a bare float. But yeah, that's a separate discussion for another time and place. Please let me know if you come up with any ideas or even fragments of ideas!

File: README.md
Comment (SGrondin) [reply README.md:1416]: Ah, good to know!

**Comment by SGrondin** [comment]
Thank you @Sudha247 and @balat for the comments! I've pushed a new commit with a few small changes based on your feedback.

File: README.md
Comment (Sudha247) [reply README.md:1283]: Fair enough. Maybe add a one-line explanation of what a domain is (and how it relates to the notion of thread in other languages)? We can also link to the manual page on parallelism https://v2.ocaml.org/manual/parallelism.html. I'm still not sure about correlating thread and domain because one domain can host many systhreads.

Comment (SGrondin) [reply README.md]: Good idea, done.

**Review by Sudha247 — APPROVED** [review]
LGTM. I'll wait until @talex5 has had a chance to review before merging.

**Comment by talex5** [comment]
It's hard to see what changes with sections being moved and edited at the same time. It seems like this is combining three things:
1. Some obvious fixes, e.g. requiring OCaml 5.1
2. Moving things around
3. Documenting the executor pool

It would be helpful to separate these, as 2 needs more thinking about. For example, the section introducing `traceln` is moved later, but all of the examples use `traceln`. The part about mocks could perhaps go a bit later, but the reason for having it at the front is to show an immediate benefit of passing the environment to `main`, as people might wonder why we're doing that.

The tracing section comes right after the fibers section so you can see visually what it looks like to run fibers concurrently. Ideally, we'd also illustrate the other examples, and readers can run the tracing tool to understand any code they write. As the tools aren't ready yet, it currently just shows how to dump the logs, which I agree isn't useful here, but this is just a place-holder while that's being written.

**Comment by talex5** [comment]
Most of this was merged in #650, so closing.

**Pull request closed without merging**
